Framework/API/inc/MantidAPI/Algorithm.h

```diff
@@ -304,12 +307,26 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
 
   void copyNonWorkspaceProperties(IAlgorithm *alg, int periodNum);
 
+  const Parallel::Communicator &communicator() const;
+  void setCommunicator(const Parallel::Communicator &communicator);
+
 protected:
   /// Virtual method - must be overridden by concrete algorithm
   virtual void init() = 0;
   /// Virtual method - must be overridden by concrete algorithm
   virtual void exec() = 0;
 
+  void exec(Parallel::ExecutionMode executionMode);
+  virtual void execDistributed();
+  virtual void execMasterOnly();
+  virtual void execNonMaster();
+
+  virtual Parallel::ExecutionMode getParallelExecutionMode(
+      const std::map<std::string, Parallel::StorageMode> &storageModes) const;
+
+  Parallel::ExecutionMode
```

Comment:
Non member sounds good. I wondered if there would be any difference for MDAlgorithms, but that's a future issue to worry about.